### DISCUSSION TITLE
Pull PSBT code out of `bitcoin` and `miniscript`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3210] [General]
Is there any reason we don't create a `github.com/rust-bitcoin/psbt-v0` repository and copy the code out of `bitcoin` and `miniscript`?

Now that everyone is sitting on `bitcoin v0.32` for a while it seems like an oportune time to do it. Then we can deprecate the original code and one day delete it.

The main thing this buys us is other folk who care more about PSBT can hack on it and it can have a different review/merge policy than `rust-bitcoin` (even a different set of maintainers if someone will step forward).

The [`psbt-v2`](https://crates.io/crates/psbt-v2) crate could then depend on it and not import it as it currently is doing.

**Comment by tcharding** [comment]
cc @DanGould

**Reply by DanGould** [reply]
This strategy makes sense to me. If we need review from people in this crate I think we can just ask them directly to come over and take a peek.

**Reply by tcharding** [reply]
Yep, most definitely.

**Comment by Kixunil** [comment]
I'd very much like to see it, though I'm not sure how to resolve coordination with `primitives`&co. Also doing so makes it harder to make pull requests which update the APIs in lockstep. This not only saves people synchronizing the crate but also makes us better informed about APIs if we're forced to also update some of the users of APIs. (e.g. BIP32 changes will update PSBT and I suspect there are things to improve because of it)

**Reply by tcharding** [reply]
We are going to need a good decent few rounds or RC testing when we release (and start using) `primitives` so I think we can just let `psbt` be part of that.

**Reply by Kixunil** [reply]
Maybe wait until primitives 1.0 is out? (or not *that* long) Though we still need `SighashCache` and PSBT is a valuable API testing resource of its API which I want to change.

**Comment by storopoli** [comment]
Why not just `psbt` instead of `v0` and we can deal with versions internally or maybe with features somehow?

I am seeing a lot of painful breakage in the future if we commit to only `v0` now.

**Reply by Kixunil** [reply]
I believe the intention was to not change too many things first and only do it later. But if the project will eventually become something better then it should be just called `psbt` from the start.

**Reply by tcharding** [reply]
I figured if the sole purpose is pulling the current code out, then being explicit that its v0 would be useful. Also, if its to be released the name `pstb-v0` is available on crates.io. And I have `psbt-v2` on crates.io already so we can use / re-use that later if needed.

Of course, plain old `psbt` would be a good name but that is not available on crates.io (I have already unsuccessfully asked the author for the name).

**Reply by Kixunil** [reply]
If you want to remove it you can just type `git rm -f bitcoin/psbt && git commit -am 'Removed PSBT'` :)

**Comment by apoelstra** [comment]
Yeah, this seems good. The historical reason is that we viewed psbtv0 as a "dumb data structure" where every role was just setting fields, except the finalizer who did all the hard work. So the idea was that we'd put the dumb stuff into rust-bitcoin where it'd be a stable data structure (lol) and people could have different finalizers -- the Miniscript one, special-purpose ones, a simple template-based one that might live directly in rust-bitcoin, etc.

Anyway it turned out that the dumb parts weren't that dumb, the stable types weren't that stable, and people who wanted to work on this struggled to get traction within the rust-bitcoin crate which is focused on other stuff right now.

**Reply by Kixunil** [reply]
Oh, you remind me what I think the overall design of a good PSBT crate should look like:

* have low-level "dumb data structure" definitions for all versions, they can be exposed for advanced users once all types are stable
* implement deserialize by deserializing into them based on the version number
* have a higher-level abstraction that gets created by converting the low-level version
* have common methods for common operations
* have special accessors for version-specific operations like this `fn as_version2(&self) -> Option<&impl Version2Operations>`
* possibly have a (proc) macro for user's types to deserialize into with field presence checking (like `serde`)

**Comment by tcharding** [comment]
Today I hacked up https://github.com/tcharding/rust-psbt-v0

My plan is to release `v0.1.0` once the [bug fix](https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/pull/16) to `rust-bitcoin-maintainer-tools` merges. The crate is then a drop in replacement for the `psbt` modules in `rust-bitcoin 0.32.2` and `rust-miniscript 12.2.0`. Next steps:

- Move the repo to github.com/rust-bitcoin/rust-psbt
- Deprecate the `psbt` modules in a point release to `rust-bitcoin` (and `rust-miniscript` if wanted)
- Delete the `psbt` module from `master`
- Migrate all PSBT relate issues to the new repo
- Get some buy in from other devs and get hacking on the new repo, some ideas:
  - Re-write the signing API  
  - Implement PSBTv2 (can pull code from my other crate `psbt-v2`)
  - Perhaps re-write the whole API (if it is a useful use of time)

**Reply by apoelstra** [reply]
This all sounds good, though I think we should be less eager to deprecate the module in the existing releases, since it's kind of a PITA for people to migrate to a whole new crate, and the existing code works and is used in production.

**Reply by tcharding** [reply]
Genuine question, do you not think that it would give us a developmental speedup here in `rust-bitcoin` if all PSBT stuff was somewhere else:
- No more issues/PRs
- Easy enough to point people to the other crate if they do raise issues
- No need to do changes to `psbt` module when we move/break other things (eg adding use of extension traits)

The downside was already pointed out by @Kixunil somewhere that the `psbt` module can be seen as a useful user of the rest of the `bitcoin` code.

**Reply by tcharding** [reply]
Also if we do not deprecate it then we are in the exact same position we are now with the `psbt-v2` crate I created - namly I have no motivation to work on a crate that is unused. Deprecating and _forcing_ development over to `rust-pstb` gives us both users of the new crate and a reason to work on it. As it is the `psbt` module is a second class citizen, I rekon we should remove all these second class citizen modules (*cough* bip32) so that what is left in `rust-bitcoin` is stuff we _really_ care about and work on. And free up others to work on the other stuff.

**Reply by tcharding** [reply]
Part of my motivation is that we are getting more ~anal~ picky in `rust-bitcoin` and there is getting to be more and more required knowledge to patch it. If we can move things with less finicky requirements to separate crates then I think this helps cultivate more devs getting involved.

**Reply by apoelstra** [reply]
All good arguments. Ok, ACK from me on deprecating the module.

**Reply by tcharding** [reply]
I've hacked up [a branch on `bdk`](https://github.com/bitcoindevkit/bdk/pull/1621) to see how hard it would be for them to use a new crate. Also to see if we get any buy in from them, this could be a heuristic for if the new crate is useful or not. (Discovered a complexity with the `miniscript::plan` module that I had not noticed too, so win!)

cc @notmandatory to represent possible buy in from `bdk` and @DanGould who is interested in PSBTs

**Reply by DanGould** [reply]
Works for me, though I have a feeling I'll be on 0.32 for a while to stay compatible with the rest of the rust projects. Could a patch release include psbt-v0 if the interface stays the same, or would that necessarily create a minor bump?

**Reply by tcharding** [reply]
Good question, this is the same question I raise here: https://github.com/bitcoindevkit/bdk/pull/1621#issuecomment-2364197120

(Note pre-1.0 a patch release is considered a minor release.)

**Comment by tcharding** [comment]
Published `rust-psbt-v0 0.1.0` - a direct import of code from `rust-bitcoin 0.32.2`'s `psbt` module. As such this crate is a drop in replacement for the `bitcoin::psbt` module. Does not include any `miniscript` stuff.

https://crates.io/crates/psbt-v0

**Comment by tcharding** [comment]
Can I move it over into the `rust-bitcoin `org @apoelstra?

**Reply by apoelstra** [reply]
Yep, go for it. Did a quick check -- it's in better shape than most things are when we initially move them into the org.

**Reply by tcharding** [reply]
Thanks for looking.

**Comment by tcharding** [comment]
As part of this project consider downstream:
- AnchorWatch
- Portal
- BDK (specifically the new `PsbtParams` stuff)

**Comment by tcharding** [comment]
@nyonson and I are starting to push on PSBT. Can I quickly clarify a few things @apoelstra before I lead him up the garden path.

I believe everything in the discussion thread so far is accurate and still relevant

The problem we are trying to solve is:

- PSBT is a higher level abstraction than most of `rust-bitcoin`
- The `psbt` module is currently a second class citizen
- (Both arguments apply also to `miniscript`)
- We want to work on PSBTv2

High level solution

- Create `psbt` crate in `rust-bitcoin` repo
- Tease out serialization logic into the new crate (as done in #3507)
- Depend on new crate for current `psbt` module. Then deprecate the module.
- [Update https://crates.io/crates/psbt-v0 so its still a drop in replacement but uses the new crate]
- Work on https://crates.io/crates/psbt-v2 for new APIs (both v0 and v2 if desired)
- Pull `minisicript` finalizer code up into `psbt-v2` crate (deprecate and/or remove from `miniscript`)

@apoelstra the `minisicript` mentions need special consideration from you please.

**Reply by apoelstra** [reply]
This all seems fine. Probably okay to keep the `psbt` stuff in `miniscript` for a while and just add a `psbt` dependency to `miniscript`. Then in parallel reimplement/copy the minscript finalizer into `psbt-v2` until it's at feature-parity and we can delete the `miniscript` one.

**Reply by tcharding** [reply]
BOOM! Thanks
